File: Assets/Bundles/Shared/Resources/TS/UserInput/Touchscreen.lua
-- Compiled with unity-ts v2.1.0-75
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local GestureDriver = require("Shared/TS/UserInput/Drivers/GestureDriver").GestureDriver
local TouchscreenDriver = require("Shared/TS/UserInput/Drivers/TouchscreenDriver").TouchscreenDriver
local Touchscreen
do
	Touchscreen = setmetatable({}, {
		__tostring = function()
			return "Touchscreen"
		end,
	})
	Touchscreen.__index = Touchscreen
	function Touchscreen.new(...)
		local self = setmetatable({}, Touchscreen)
		return self:constructor(...) or self
	end
	function Touchscreen:constructor()
		self.bin = Bin.new()
		self.touchscreenDriver = TouchscreenDriver:instance()
		self.gestureDriver = GestureDriver.new()
		self.Touch = Signal.new()
		self.TouchTap = Signal.new()
		self.PrimaryTouch = Signal.new()
		self.PrimaryTouchTap = Signal.new()
		self.Pan = Signal.new()
		self.Pinch = Signal.new()
		self.bin:Add(self.Touch)
		self.bin:Add(self.TouchTap)
		self.bin:Add(self.PrimaryTouch)
		self.bin:Add(self.PrimaryTouchTap)
		self.bin:Add(self.Pan)
		self.bin:Add(self.Pinch)
		self.bin:Connect(self.touchscreenDriver.Touch, function(touchIndex, position, phase)
			self.Touch:Fire(touchIndex, position, phase)
			if touchIndex == 0 then
				self.PrimaryTouch:Fire(position, phase)
			end
		end)
		self.bin:Connect(self.touchscreenDriver.TouchTap, function(touchIndex, position, phase)
			self.TouchTap:Fire(touchIndex, position, phase)
			if touchIndex == 0 then
				self.PrimaryTouchTap:Fire(position, phase)
			end
		end)
		self.bin:Add(self.gestureDriver)
		self.bin:Add(self.gestureDriver.Pan:Proxy(self.Pan))
		self.bin:Add(self.gestureDriver.Pinch:Proxy(self.Pinch))
	end
	function Touchscreen:Destroy()
		self.bin:Destroy()
	end
end
return {
	Touchscreen = Touchscreen,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/UserInput/Touchscreen.ts
import { Bin } from "Shared/Util/Bin";
import { Signal } from "Shared/Util/Signal";
import { GestureDriver } from "./Drivers/GestureDriver";
import { TouchscreenDriver } from "./Drivers/TouchscreenDriver";

export class Touchscreen {
	private readonly bin = new Bin();
	private readonly touchscreenDriver = TouchscreenDriver.Instance();
	private readonly gestureDriver = new GestureDriver();

	public readonly touch = new Signal<[touchIndex: number, position: Vector3, phase: TouchPhase]>();
	public readonly touchTap = new Signal<[touchIndex: number, position: Vector3, phase: InputActionPhase]>();

	public readonly primaryTouch = new Signal<[position: Vector3, phase: TouchPhase]>();
	public readonly primaryTouchTap = new Signal<[position: Vector3, phase: InputActionPhase]>();

	public readonly pan = new Signal<[position: Vector3, phase: TouchPhase]>();
	public readonly pinch = new Signal<[distance: number, scale: number, phase: TouchPhase]>();

	constructor() {
		this.bin.Add(this.touch);
		this.bin.Add(this.touchTap);
		this.bin.Add(this.primaryTouch);
		this.bin.Add(this.primaryTouchTap);
		this.bin.Add(this.pan);
		this.bin.Add(this.pinch);
		this.bin.Connect(this.touchscreenDriver.touch, (touchIndex, position, phase) => {
			this.touch.Fire(touchIndex, position, phase);
			if (touchIndex === 0) {
				this.primaryTouch.Fire(position, phase);
			}
		});
		this.bin.Connect(this.touchscreenDriver.touchTap, (touchIndex, position, phase) => {
			this.touchTap.Fire(touchIndex, position, phase);
			if (touchIndex === 0) {
				this.primaryTouchTap.Fire(position, phase);
			}
		});
		this.bin.Add(this.gestureDriver);
		this.bin.Add(this.gestureDriver.pan.Proxy(this.pan));
		this.bin.Add(this.gestureDriver.pinch.Proxy(this.pinch));
	}

	/**
	 * Cleans up the touchscreen listener.
	 */
	public Destroy() {
		this.bin.Destroy();
	}
}
